nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/nar/StandardNarManager.java

```diff
@@ -0,0 +1,504 @@
+        this.narLoader = narLoader;
+        this.webClientService = webClientService;
+        this.sslContext = sslContext;
+        this.installExecutorService = Executors.newSingleThreadExecutor();
+        this.deleteExecutorService = Executors.newSingleThreadExecutor();
+    }
+
+    // This serves two purposes...
+    // 1. Any previously stored NARs need to have their extensions loaded and made available for use during start up since they won't be in any of the standard NAR directories
+    // 2. NarLoader keeps track of NARs that were missing dependencies to consider them on future loads, so this restores state that may have been lost on a restart
+    @Override
+    public void afterPropertiesSet() throws IOException {
+        final Collection<NarPersistenceInfo> narInfos = loadExistingNars();
+        restoreState(narInfos);
+    }
+
+    private Collection<NarPersistenceInfo> loadExistingNars() throws IOException {
+        final Collection<NarPersistenceInfo> narInfos = persistenceProvider.getAllNarInfo();
+        LOGGER.info("Initializing NAR Manager, loading {} previously stored NARs", narInfos.size());
+
+        final Collection<File> narFiles = narInfos.stream()
+                .map(NarPersistenceInfo::getNarFile)
+                .collect(Collectors.toSet());
+        narLoader.load(narFiles);
+        return narInfos;
+    }
+
+    private void restoreState(final Collection<NarPersistenceInfo> narInfos) {
+        for (final NarPersistenceInfo narInfo : narInfos) {
+            try {
+                final NarNode narNode = restoreNarNode(narInfo);
+                narNodesById.put(narNode.getIdentifier(), narNode);
+                LOGGER.debug("Restored NAR [{}] with state [{}] and identifier [{}]",
+                        narNode.getManifest().getCoordinate(), narNode.getState(), narNode.getIdentifier());
+            } catch (final Exception e) {
+                LOGGER.warn("Failed to restore NAR for [{}]", narInfo.getNarFile().getAbsolutePath(), e);
+            }
+        }
+    }
+
+    private NarNode restoreNarNode(final NarPersistenceInfo narInfo) throws IOException {
+        final File narFile = narInfo.getNarFile();
+        final NarManifest manifest = NarManifest.fromNarFile(narFile);
+        final BundleCoordinate coordinate = manifest.getCoordinate();
+        final String identifier = createIdentifier(coordinate);
+        final NarState state = determineNarState(manifest);
+        final String narDigest = computeNarDigest(narFile);
+
+        return NarNode.builder()
+                .identifier(identifier)
+                .narFile(narFile)
+                .narFileHexDigest(narDigest)
+                .manifest(manifest)
+                .source(NarSource.valueOf(narInfo.getNarProperties().getSourceType()))
+                .sourceIdentifier(narInfo.getNarProperties().getSourceId())
+                .state(state)
+                .build();
+    }
+
+    @Override
+    public void destroy() {
```

Comment:
Instead of implementing `DisposableBean`, this could implement `Closeable` and use the `close()` method.